liquidity/liquidity.go

```diff
@@ -1020,6 +1050,123 @@ func (m *Manager) currentSwapTraffic(loopOut []*loopdb.LoopOut,
 	return traffic
 }
 
+// dispatchStickyLoopOut attempts to dispatch a loop out swap that will
+// automatically retry its execution with an amount based backoff.
+func (m *Manager) dispatchStickyLoopOut(ctx context.Context,
+	out loop.OutRequest, retryCount uint16, amountBackoff float64) error {
+
+	for i := 0; i < int(retryCount); i++ {
+		// Dispatch the swap.
+		swap, err := m.cfg.LoopOut(ctx, &out)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("loop out automatically dispatched: hash: %v, "+
+			"address: %v, amount %v", swap.SwapHash,
+			swap.HtlcAddress, out.Amount)
+
+		updates := make(chan *loopdb.SwapState)
+
+		// Monitor the swap state.
+		go m.waitForSwapPayment(
+			ctx, swap.SwapHash, updates, defaultPaymentWaitTimeout,
+		)
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		case update := <-updates:
+			if update == nil {
+				// If update is nil then no update occurred
+				// within the defined timeout period. It's
+				// better to return and not attempt a retry.
+
+				log.Infof("No update received for swap")
+
+				return nil
+			}
+
+			if *update == loopdb.StateFailOffchainPayments {
+				// If we failed to pay the server, we will
+				// decrease the amount of the swap and try
+				// again.
+				out.Amount -= btcutil.Amount(
+					float64(out.Amount) * amountBackoff,
+				)
+
+				log.Infof("swap %v: amount backoff new amount="+
+					"%v", swap.SwapHash, out.Amount)
+
+				continue
+			} else {
+				// If the update channel did not return an
+				// off-chain payment failure we won't retry.
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+// waitForSwapPayment waits for a swap to progress beyond the stage of
+// forwarding the payment to the server through the network. It returns the
+// final update on the outcome through a channel.
+func (m *Manager) waitForSwapPayment(ctx context.Context, swapHash lntypes.Hash,
+	updateChan chan *loopdb.SwapState, timeout time.Duration) {
+
+	startTime := time.Now()
+	var (
+		swap     *loopdb.LoopOut
+		err      error
+		interval time.Duration
+	)
+
+	if m.params.CustomPaymentCheckInterval != 0 {
+		interval = m.params.CustomPaymentCheckInterval
+	} else {
+		interval = defaultPaymentCheckInterval
+	}
+
+	for time.Since(startTime) < timeout {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+		}
+
+		swap, err = m.cfg.GetLoopOut(swapHash)
+		if err != nil {
+			log.Errorf("Error getting swap: %v", err)
+			return
+		}
+
+		// If no update has occurred yet, continue in order to wait.
+		update := swap.LastUpdate()
+		if update == nil {
+			continue
+		}
+
+		// If the swap is still pending, we will continue to wait.
+		switch update.State {
+		case loopdb.StateFailOffchainPayments:
+			updateChan <- &update.State
+			return
+
+		case loopdb.StateHtlcPublished:
```

Comment:
Loop out swaps won't reach this state as it's used for loop ins.

Reply:
Right, was confused with the `SERVER_HTLC_PUBLISHED` state, which is used for loop out.